backend/core/src/Core.API/ResponseHandling/SignatureVerificationMiddleware.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright 2023 Quantoz Technology B.V. and contributors. Licensed
+// under the Apache License, Version 2.0. See the NOTICE file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+using Core.Domain.Exceptions;
+using Core.Presentation.Models;
+using Newtonsoft.Json;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Core.API.ResponseHandling
+{
+    public class SignatureVerificationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger<SignatureVerificationMiddleware> _logger;
+
+        public SignatureVerificationMiddleware(
+            RequestDelegate next,
+            ILogger<SignatureVerificationMiddleware> logger)
+        {
+            _next = next;
+            _logger = logger;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            try
+            {
+                // Retrieve headers from the request
+                string? publicKeyHeader = context.Request.Headers["x-public-key"];
+                string? signatureHeader = context.Request.Headers["x-signature"];
+
+                // Retrieve method from the request
+                var method = context.Request.Method;
+
+                // Ensure both headers are present
+                if (string.IsNullOrWhiteSpace(publicKeyHeader) || string.IsNullOrWhiteSpace(signatureHeader))
+                {
+                    _logger.LogError("Missing x-signature header");
+                    var customErrors = new CustomErrors(new CustomError("Forbidden", "Missing Header", "x-signature"));
+                    await WriteCustomErrors(context.Response, customErrors, (int)HttpStatusCode.Forbidden);
+                    return;
+                }
+
+                // Get the current Unix UTC timestamp (rounded to 30 seconds)
+                long currentTimestamp = (long)(DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalSeconds;
+                currentTimestamp = (currentTimestamp / 30) * 30; // Round to the nearest 30 seconds
+
+                string? postPayload = string.Empty;
+                string? payload = null;
+
+                switch (method)
+                {
+                    case "POST":
+                    case "PUT":
+                        {
+                            using (StreamReader reader = new(context.Request.Body, Encoding.UTF8))
+                            {
+                                postPayload = await reader.ReadToEndAsync();
+                            }
+
+                            payload = JsonConvert.SerializeObject(new
```

Comment:
I think we shouldn't make this an anonymous object, but rather a defined one so it's clear what the app should use. You currently rely on the app and you using the same postPayload, so if the app would use payload you'd suddenly have a mismatch.

Reply:
done